borg/archiver.py

```diff
@@ -353,6 +355,155 @@ def do_extract(self, args):
                 self.print_warning("Include pattern '%s' never matched.", pattern)
         return self.exit_code
 
+    def do_diff(self, args):
+        """Diff archive contents"""
+        def dump_json(dictionary):
+            if dump_json.wants_comma:
+                sys.stdout.write(",")
+            dump_json.wants_comma = True
+            json.dump(dictionary, sys.stdout)
+        dump_json.wants_comma = False
+
+        def format_bytes(count):
+            if count is None:
+                return "<deleted>"
+            return format_file_size(count)
+
+        def compare_chunks(lhs, rhs, lhs_archive, rhs_archive):
+            lhs_hash = sha256()
+            rhs_hash = sha256()
+            for chunk in lhs_archive.pipeline.fetch_many(lhs):
+                lhs_hash.update(chunk)
+            for chunk in rhs_archive.pipeline.fetch_many(rhs):
+                rhs_hash.update(chunk)
+            return lhs_hash.digest() == rhs_hash.digest()
```

Comment:
also, instead of hardcoding sha256, rather do it like that:
def compare_chunks(..., hash=hashlib.sha512):
    hash1 = hash()
    hash2 = hash()

Note that sha512 is faster than sha256 on 64bit machines.